File: biocentral_api/python/biocentral_api/clients/custom_models_client.py
from tqdm import tqdm
from typing import Dict, Any, List

from .tasks import BiocentralServerTask, DTOHandler

from .._generated import ApiClient, StartTrainingRequest, ConfigVerificationRequest, CustomModelsApi, \
    SequenceTrainingData, TaskDTO, TaskStatus, StartInferenceRequest


class _TrainingDTOHandler(DTOHandler):

    def handle(self, dtos: List[TaskDTO]):
        for dto in dtos:
            status = dto.status
            if status == TaskStatus.FINISHED:
                # TODO Error handling
                return dto.biotrainer_result
        return None

    def update_tqdm(self, dtos: List[TaskDTO], pbar: tqdm) -> tqdm:
        for dto in dtos:
            status = dto.status
            if status == TaskStatus.RUNNING and dto.biotrainer_update is not None:
                training_iteration = dto.biotrainer_update.training_iteration
                if training_iteration is not None:
                    epoch = training_iteration[1]['epoch']
                    pbar.set_description(f"Training epoch {epoch}")
                    pbar.update(1)  # TODO Improve handling of progress bar updates
            if status == TaskStatus.FINISHED:
                pbar.set_description(f"Finished training")
                pbar.close()
                break
            if status == TaskStatus.FAILED:
                pbar.set_description(f"Training failed")
                pbar.close()
                break
        return pbar


class _InferenceDTOHandler(DTOHandler):

    def handle(self, dtos: List[TaskDTO]):
        for dto in dtos:
            status = dto.status
            if status == TaskStatus.FINISHED:
                # TODO Error handling
                return dto.predictions
        return None

    def update_tqdm(self, dtos: List[TaskDTO], pbar: tqdm) -> tqdm:
        for dto in dtos:
            status = dto.status
            match status:
                case TaskStatus.PENDING:
                    pbar.set_description(f"Waiting for inference to start..")
                case TaskStatus.RUNNING:
                    pbar.set_description(f"Running inference..")
                case TaskStatus.FINISHED:
                    pbar.set_description(f"Finished inference!")
                    pbar.close()
                    break
                case TaskStatus.FAILED:
                    pbar.set_description(f"Inference failed!")
                    pbar.close()
                    break
        return pbar


class CustomModelsClient:
    def train(self, api_client: ApiClient, config: Dict[str, Any],
              training_data: List[SequenceTrainingData]) -> BiocentralServerTask:
        custom_models_api = CustomModelsApi(api_client)
        config_verification_request = ConfigVerificationRequest(config_dict=config)

        try:
            config_verification_response = custom_models_api.verify_config_api_v1_custom_models_service_verify_config_post(
                config_verification_request
            )
            config_error = config_verification_response.error
            if config_error:
                raise Exception(f"Config verification failed: {config_error}")
        except Exception as e:
            raise e

        training_dto_handler = _TrainingDTOHandler()

        start_training_request = StartTrainingRequest(config_dict=config, training_data=training_data)
        task_id = custom_models_api.start_training_api_v1_custom_models_service_start_training_post(
            start_training_request).task_id
        return BiocentralServerTask(task_id=task_id, api_client=api_client, dto_handler=training_dto_handler)

    def inference(self, api_client: ApiClient, model_hash: str, inference_data: Dict[str, str]):
        custom_models_api = CustomModelsApi(api_client)
        start_inference_request = StartInferenceRequest(model_hash=model_hash, sequence_data=inference_data)
        task_id = custom_models_api.start_inference_api_v1_custom_models_service_start_inference_post(
            start_inference_request).task_id
        return BiocentralServerTask(task_id=task_id, api_client=api_client, dto_handler=_InferenceDTOHandler())

File: biocentral_api/dart/lib/src/clients/custom_models_client.dart
import 'package:biocentral_api/src/api.dart' as gen;
import 'package:biocentral_api/src/model/config_options_response.dart';
import 'package:biocentral_api/src/model/config_verification_request.dart';
import 'package:biocentral_api/src/model/config_verification_response.dart';
import 'package:biocentral_api/src/model/prediction.dart';
import 'package:biocentral_api/src/model/biotrainer_model_result.dart';
import 'package:biocentral_api/src/model/sequence_data.dart';
import 'package:biocentral_api/src/model/start_inference_request.dart';
import 'package:biocentral_api/src/model/start_training_request.dart';
import 'package:biocentral_api/src/model/task_dto.dart';
import 'package:biocentral_api/src/model/task_status.dart';
import 'package:built_collection/built_collection.dart';
import 'package:built_value/json_object.dart';

import 'tasks/biocentral_server_task.dart';
import 'tasks/dto_handler.dart';
import 'tasks/submit_task.dart';

class _TrainingDtoHandler extends DtoHandler<BiotrainerModelResult> {

  @override
  BiotrainerModelResult? handle(List<TaskDTO> dtos) {
    for (final dto in dtos) {
      if (dto.status == TaskStatus.FINISHED) {
        return dto.biotrainerResult;
      }
    }
    return null;
  }
}

class _InferenceDtoHandler extends DtoHandler<Map<String, List<Prediction>>> {
  @override
  Map<String, List<Prediction>>? handle(List<TaskDTO> dtos) {
    for (final dto in dtos) {
      if (dto.status == TaskStatus.FINISHED) {
        return dto.predictions?.toMap().map((k, v) => MapEntry(k, v.toList()));
      }
    }
    return null;
  }
}

class CustomModelsClient {

  Future<List<dynamic>?> getConfigOptionsForProtocol({required gen.BiocentralApi api, required String protocol}) async {
    final cmApi = api.getCustomModelsApi();
    final resp = await cmApi.configOptionsApiV1CustomModelsServiceConfigOptionsProtocolGet(protocol: protocol);
    final ConfigOptionsResponse optionsResponse = resp.data!;
    return optionsResponse.options.toList();
  }

  Future<String?> verifyTrainingConfig({
    required gen.BiocentralApi api,
    required Map<String, dynamic> config,
  }) async {
    final cmApi = api.getCustomModelsApi();

    final verifyReq =
        ConfigVerificationRequest((b) => b..configDict.replace(config.map((k, v) => MapEntry(k, JsonObject(v)))));
    final verifyResp = await cmApi.verifyConfigApiV1CustomModelsServiceVerifyConfigPost(
      configVerificationRequest: verifyReq,
    );
    final ConfigVerificationResponse verification = verifyResp.data!;
    String? err = verification.error;
    err = (err?.isEmpty ?? true) ? null : err;
    return err;
  }

  /// Verify config then start training.
  Future<BiocentralServerTask<BiotrainerModelResult?>> train({
    required gen.BiocentralApi api,
    required Map<String, dynamic> config,
    required List<SequenceData> trainingData,
  }) async {
    final cmApi = api.getCustomModelsApi();

    // Verify config
    final configVerificationError = await verifyTrainingConfig(api: api, config: config);

    if (configVerificationError != null) {
      throw Exception(configVerificationError);
    }

    final handler = _TrainingDtoHandler();
    final startReq = StartTrainingRequest((b) => b
      ..configDict.replace(config.map((k, v) => MapEntry(k, JsonObject(v))))
      ..trainingData.replace(BuiltList<SequenceData>(trainingData)));

    final taskId = await submitTask(() => cmApi
        .startTrainingApiV1CustomModelsServiceStartTrainingPost(startTrainingRequest: startReq));
    return BiocentralServerTask<BiotrainerModelResult>(taskId: taskId, api: api, dtoHandler: handler);
  }

  /// Start inference for a trained model.
  Future<BiocentralServerTask<Map<String, List<Prediction>>>> inference({
    required gen.BiocentralApi api,
    required String modelHash,
    required Map<String, String> sequenceData,
  }) async {
    final cmApi = api.getCustomModelsApi();
    final startReq = StartInferenceRequest((b) => b
      ..modelHash = modelHash
      ..sequenceData.replace(BuiltMap<String, String>(sequenceData)));

    final taskId = await submitTask(() => cmApi
        .startInferenceApiV1CustomModelsServiceStartInferencePost(startInferenceRequest: startReq));
    return BiocentralServerTask<Map<String, List<Prediction>>>(
      taskId: taskId,
      api: api,
      dtoHandler: _InferenceDtoHandler(),
    );
  }
}
